### PR TITLE
[SYCL-MLIR] Support opaque pointers in cgeist LLVM IR output

### PR DESCRIPTION
**Pull request by sommerlukas**
Next step to add opaque pointer support for the SYCL-MLIR project: Enable `cgeist` to produce LLVM IR using opaque pointers as output from the compilation. 

Emitting LLVM IR with typed or opaque pointers is controlled by the `--use-opaque-pointers` option. 

Amongst others, the following steps were taken to enable this:
* Detect recursive structs based on Clang types, not LLVM types resulting from translation
* Fix some issues in the Polygeist and SYCL to LLVM lowerings
* Translate type attributes of `getelementptr` and `alloca` LLVM dialect operations

### CODE REVIEW

**Comment by sommerlukas** [comment]
Running the SYCL end-to-end tests with `cgeist` and opaque pointers enabled was also tested: 
```
  Unsupported      : 749
  Passed           : 228
  Expectedly Failed: 245
  Timed Out        :   1
  Failed           : 104
```

The failures are mainly caused by issues:
* The lack of support for the LLVM `TargetExtType`, used to represent SPIR-V image types, in the MLIR LLVM dialect, which will be added in a separate PR. This affects one of the failing test cases.
* The remaining failures are caused by a `segfault` in the OpenCL CPU driver when consuming the device SPIR-V and translating it to LLVM IR. So compilation with `cgeist` works for all these cases without error, but the driver seems unable to consume the SPIR-V produced from LLVM IR using opaque pointers.

Together, these issues mean we can't yet change the default of `cgeist` to use opaque pointers by default.

**Comment by etiotto** [comment]
> Running the SYCL end-to-end tests with `cgeist` and opaque pointers enabled was also tested:
> 
> ```
>   Unsupported      : 749
>   Passed           : 228
>   Expectedly Failed: 245
>   Timed Out        :   1
>   Failed           : 104
> ```
> 
> The failures are mainly caused by issues:
> 
> * The lack of support for the LLVM `TargetExtType`, used to represent SPIR-V image types, in the MLIR LLVM dialect, which will be added in a separate PR. This affects one of the failing test cases.
> * The remaining failures are caused by a `segfault` in the OpenCL CPU driver when consuming the device SPIR-V and translating it to LLVM IR. So compilation with `cgeist` works for all these cases without error, but the driver seems unable to consume the SPIR-V produced from LLVM IR using opaque pointers.
> 
> Together, these issues mean we can't yet change the default of `cgeist` to use opaque pointers by default.

This is good progress. For the 2nd problem: comparing the LLVM IR produced by DPC++ vs the one produced by cgeist may expose the difference that is causing the segfault.

File: polygeist/tools/cgeist/Test/Verification/sycl/builtins.cpp [hunk 1]
```diff
@@ -1,5 +1,5 @@
 // RUN: clang++ -fsycl -fsycl-device-only -fsycl-targets=spir64-unknown-unknown-syclmlir -O0 -w -emit-mlir -o - %s | FileCheck %s
-// RUN: clang++ -fsycl -fsycl-device-only -fsycl-targets=spir64-unknown-unknown-syclmlir -O0 -w -emit-llvm -S -o %t %s && rm %t
+// RUN: clang++ -fsycl -fsycl-device-only -fsycl-targets=spir64-unknown-unknown-syclmlir -O1 -w -emit-llvm -S -o %t %s && rm %t
```
Comment (etiotto) [inline polygeist/tools/cgeist/Test/Verification/sycl/builtins.cpp:2]: Did -O0 not work?

File: polygeist/tools/cgeist/driver.cc [hunk 2]
```diff
@@ -387,7 +387,7 @@ static LogicalResult optimize(mlir::MLIRContext &Ctx,
   if (OptLevel != llvm::OptimizationLevel::O0) {
     PM.addPass(polygeist::createArgumentPromotionPass());
     PM.addPass(polygeist::createKernelDisjointSpecializationPass(
-        {options.getCgeistOpts().getRelaxedAliasing()}));
+        {options.getCgeistOpts().getRelaxedAliasing(), UseOpaquePointers}));
```
Comment (etiotto) [inline polygeist/tools/cgeist/driver.cc:390]: We could create a variable holding `{options.getCgeistOpts().getRelaxedAliasing(), UseOpaquePointers})` rather than repeating the same options every time we add a pass.

File: mlir-sycl/lib/Conversion/SYCLToLLVM/DPCPP.cpp [hunk 3]
```diff
@@ -2186,7 +2200,7 @@ class GroupGetLocalLinearIDPattern
     rewriter.create<memref::StoreOp>(loc, localIDVal, localID);
     const auto group = adaptor.getGroup();
     const auto convGroupTy =
-        typeConverter->convertType(op.getGroup().getType());
+        typeConverter->convertType(op.getGroup().getType().getElementType());
```
Comment (etiotto) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/DPCPP.cpp:2203]: Was this a problem unrelated to the use of opaque ptrs ?

File: polygeist/test/polygeist-opt/bareptrlowering.mlir [hunk 4]
```diff
@@ -461,7 +461,7 @@ func.func private @subindexop_memref_same_dim(%arg0: memref<4x4xf32>, %arg1: ind
 // CHECK-SAME:                                        %[[VAL_0:.*]]: !llvm.ptr) -> !llvm.ptr
 // CHECK-NEXT:      %[[VAL_1:.*]] = llvm.mlir.constant(0 : index) : i64
 // CHECK-NEXT:      %[[VAL_2:.*]] = llvm.mlir.constant(0 : i64) : i64
-// CHECK-NEXT:      %[[VAL_3:.*]] = llvm.getelementptr %[[VAL_0]]{{\[}}%[[VAL_2]], %[[VAL_1]]] : (!llvm.ptr, i64, i64) -> !llvm.ptr, f32
+// CHECK-NEXT:      %[[VAL_3:.*]] = llvm.getelementptr %[[VAL_0]]{{\[}}%[[VAL_2]], 0] : (!llvm.ptr, i64) -> !llvm.ptr, !llvm.struct<(f32)>
```
Comment (etiotto) [inline polygeist/test/polygeist-opt/bareptrlowering.mlir:464]: VAL_1 is no longer used, so it would be nice if the compiler was not generating the operation (at line 462). If that is difficult then I would adjust the test to remove the CHECK-NEXT at line 462. Same comment for similar patterns in this file.

File: mlir-sycl/test/Conversion/SYCLToLLVM/sycl-grid-ops-to-llvm.mlir [hunk 5]
```diff
@@ -34,7 +34,7 @@ module attributes {gpu.container_module} {
     // CHECK-DAG:         %[[VAL_8:.*]] = llvm.mlir.constant(0 : i32) : i32
     // CHECK-DAG:         %[[VAL_9:.*]] = llvm.mlir.constant(0 : i32) : i32
     // CHECK-NEXT:        %[[VAL_10:.*]] = llvm.extractelement %[[VAL_1]]{{\[}}%[[VAL_9]] : i32] : vector<3xi64>
-    // CHECK-NEXT:        %[[VAL_11:.*]] = llvm.getelementptr inbounds %[[VAL_6]][0, 0, 0, %[[VAL_8]]] : (!llvm.ptr, i32) -> !llvm.ptr, i64
+    // CHECK-NEXT:        %[[VAL_11:.*]] = llvm.getelementptr inbounds %[[VAL_6]][0, 0, 0, %[[VAL_8]]] : (!llvm.ptr, i32) -> !llvm.ptr, !llvm.struct<"class.sycl::_V1::range.1", {{.*}}>
```
Comment (etiotto) [inline mlir-sycl/test/Conversion/SYCLToLLVM/sycl-grid-ops-to-llvm.mlir:37]: How come the result type has changed in several tests ?

File: polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp [hunk 6]
```diff
@@ -640,8 +640,8 @@ struct SubIndexBarePtrOpLowering : public BaseSubIndexOpLowering {
     // polygeist.subindex operation should be a memref of the element type of
     // the struct.
 
-    rewriter.replaceOpWithNewOp<LLVM::GEPOp>(subViewOp, resType,
-                                             convViewElemType, target, indices);
+    rewriter.replaceOpWithNewOp<LLVM::GEPOp>(
+        subViewOp, resType, convSourceElemType, target, indices);
```
Comment (etiotto) [inline polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp:644]: This change looks unrelated to opaque ptr.

File: polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp [hunk 7]
```diff
@@ -1634,8 +1657,43 @@ struct ConvertPolygeistToLLVMPass
...
+        // translated to the corresponding LLVM types, for example the element
+        // type attribute of GEP or alloca.
+        std::optional<bool> noTyAttrConversion;
+        if (useOpaquePointers)
+          noTyAttrConversion = std::transform_reduce(
+              op->getAttrs().begin(), op->getAttrs().end(),
+              std::optional<bool>{true},
+              [](std::optional<bool> b1,
+                 std::optional<bool> b2) -> std::optional<bool> {
+                if (!b1.has_value() || !b2.has_value()) {
```
Comment (etiotto) [inline polygeist/lib/Conversion/PolygeistToLLVM/PolygeistToLLVM.cpp:1671]: [nit]: single stmt here is braced while the single stmt at line 1665 is not braced. I would do the opposite.

File: polygeist/test/polygeist-opt/bareptrlowering.mlir [hunk 8]
```diff
@@ -477,7 +477,7 @@ func.func private @subindexop_memref_struct(%arg0: memref<4x!llvm.struct<(f32)>>
 // CHECK-SAME:                                               %[[VAL_0:.*]]: !llvm.ptr) -> !llvm.ptr
 // CHECK-NEXT:      %[[VAL_1:.*]] = llvm.mlir.constant(0 : index) : i64
 // CHECK-NEXT:      %[[VAL_2:.*]] = llvm.mlir.constant(0 : i64) : i64
-// CHECK-NEXT:      %[[VAL_3:.*]] = llvm.getelementptr %[[VAL_0]]{{\[}}%[[VAL_2]], %[[VAL_2]], %[[VAL_1]]] : (!llvm.ptr, i64, i64, i64) -> !llvm.ptr, f32
+// CHECK-NEXT:      %[[VAL_3:.*]] = llvm.getelementptr %[[VAL_0]]{{\[}}%[[VAL_2]], 0, 0] : (!llvm.ptr, i64) -> !llvm.ptr, !llvm.struct<(struct<(f32)>)>
```
Comment (etiotto) [inline polygeist/test/polygeist-opt/bareptrlowering.mlir:480]: The result has changed: i64 -> !llvm.struct<(struct<(f32)>)>. What is the reason ?

File: mlir-sycl/lib/Conversion/SYCLToLLVM/DPCPP.cpp [hunk 9]
```diff
@@ -2186,7 +2200,7 @@ class GroupGetLocalLinearIDPattern
     rewriter.create<memref::StoreOp>(loc, localIDVal, localID);
     const auto group = adaptor.getGroup();
     const auto convGroupTy =
-        typeConverter->convertType(op.getGroup().getType());
```
Comment (whitneywhtsang) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/DPCPP.cpp:2189]: Did we have a bug before?

File: polygeist/tools/cgeist/Lib/TypeUtils.cc [hunk 10]
```diff
@@ -43,6 +43,88 @@ bool isRecursiveStruct(Type *T, Type *Meta, SmallPtrSetImpl<Type *> &Seen) {
   return false;
 }
 
+bool isRecursiveStruct(
+    clang::QualType QT,
+    SmallPtrSetImpl<const clang::Type *> &RecordEncountered) {
+  if (const auto *ET = dyn_cast<clang::ElaboratedType>(QT))
```
[the remark this answers is not in the document]
Comment (sommerlukas) [reply polygeist/tools/cgeist/Lib/TypeUtils.cc:49]: I copied the `if` structure from `CodeGenTypes::getMLIRType` to make sure I do not miss any case.

**Comment by sommerlukas** [comment]
> > Running the SYCL end-to-end tests with `cgeist` and opaque pointers enabled was also tested:
> > ```
> >   Unsupported      : 749
> >   Passed           : 228
> >   Expectedly Failed: 245
> >   Timed Out        :   1
> >   Failed           : 104
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The failures are mainly caused by issues:
> > 
> > * The lack of support for the LLVM `TargetExtType`, used to represent SPIR-V image types, in the MLIR LLVM dialect, which will be added in a separate PR. This affects one of the failing test cases.
> > * The remaining failures are caused by a `segfault` in the OpenCL CPU driver when consuming the device SPIR-V and translating it to LLVM IR. So compilation with `cgeist` works for all these cases without error, but the driver seems unable to consume the SPIR-V produced from LLVM IR using opaque pointers.
> > 
> > Together, these issues mean we can't yet change the default of `cgeist` to use opaque pointers by default.
> 
> This is good progress. For the 2nd problem: comparing the LLVM IR produced by DPC++ vs the one produced by cgeist may expose the difference that is causing the segfault.

Yes, I did that and a few other things to debug the problem. 

There are some differences between the LLVM IR generated by `cgeist` for typed vs. opaque pointers. 
This is the typed pointer output:
```llvm
define weak_odr spir_kernel void @_ZTSN4sycl3_V16detail19__pf_kernel_wrapperIZZ12parallel_forRSt5arrayIiLm8EEENKUlRNS0_7handlerEE_clES7_E19kernel_parallel_forEE(%"class.sycl::_V1::range.1"* noundef byval(%"class.sycl::_V1::range.1") align 8 %0, i32 addrspace(1)* noundef align 4 %1, %"class.sycl::_V1::range.1"* noundef byval(%"class.sycl::_V1::range.1") align 8 %2, %"class.sycl::_V1::range.1"* noundef byval(%"class.sycl::_V1::range.1") align 8 %3, %"class.sycl::_V1::id.1"* noundef byval(%"class.sycl::_V1::id.1") align 8 %4) local_unnamed_addr #0 {
  %6 = addrspacecast %"class.sycl::_V1::range.1"* %0 to %"class.sycl::_V1::range.1" addrspace(4)*
  %7 = getelementptr inbounds %"class.sycl::_V1::range.1", %"class.sycl::_V1::range.1" addrspace(4)* %6, i64 0, i32 0, i32 0, i64 0
  %8 = load i64, i64 addrspace(4)* %7, align 8
  %9 = load i64, i64 addrspace(1)* getelementptr inbounds (<3 x i64>, <3 x i64> addrspace(1)* @__spirv_BuiltInGlobalInvocationId, i64 0, i64 0), align 32
  %10 = icmp slt i64 %9, %8
  br i1 %10, label %11, label %19
```

And this is the opaque pointer output: 
```llvm
define weak_odr spir_kernel void @_ZTSN4sycl3_V16detail19__pf_kernel_wrapperIZZ12parallel_forRSt5arrayIiLm8EEENKUlRNS0_7handlerEE_clES7_E19kernel_parallel_forEE(ptr noundef byval(%"class.sycl::_V1::range.1") align 8 %0, ptr addrspace(1) noundef align 4 %1, ptr noundef byval(%"class.sycl::_V1::range.1") align 8 %2, ptr noundef byval(%"class.sycl::_V1::range.1") align 8 %3, ptr noundef byval(%"class.sycl::_V1::id.1") align 8 %4) local_unnamed_addr #0 {
  %6 = addrspacecast ptr %0 to ptr addrspace(4)
  %7 = load i64, ptr addrspace(4) %6, align 8
  %8 = load i64, ptr addrspace(1) @__spirv_BuiltInGlobalInvocationId, align 32
  %9 = icmp slt i64 %8, %7
  br i1 %9, label %10, label %17
```

The most notable difference is the absence of `getelementptr` in the opaque pointer case. 

However, the LLVM IR for the opaque pointer case is still legal LLVM IR, it passes the LLVM verifier. 

It also correctly round-trips through the LLVM IR -> SPIR-V -> LLVM IR conversion, without any error. The converted SPIR-V also verifies with `spirv-val`, so this may be a problem with the driver, e.g., using an older version of the LLVM-SPIR-V translator.

File: polygeist/tools/cgeist/Lib/TypeUtils.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Lib/TypeUtils.cc:49]: We can still change it to a `TypeSwitch` now that it is copied over, or do you prefer to change both `isRecursiveStruct` and `getMLIRType` in a different PR?

File: mlir-sycl/lib/Conversion/SYCLToLLVM/DPCPP.cpp (hunk 3 above)
Comment (etiotto) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/DPCPP.cpp:2203]: ok

**Review by etiotto — APPROVED** [review]

File: polygeist/test/polygeist-opt/bareptrlowering.mlir (hunk 4 above)
Comment (sommerlukas) [reply polygeist/test/polygeist-opt/bareptrlowering.mlir:464]: These constants are removed by the canonicalization pass. Removing the `CHECK-NEXT` line for them would mean we might miss other operations that are not supposed to be there (because the line for `VAL_2` would need to be `CHECK` instead of `CHECK-NEXT`).

File: polygeist/tools/cgeist/Lib/TypeUtils.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (sommerlukas) [reply polygeist/tools/cgeist/Lib/TypeUtils.cc]: I replaced this with `TypeSwitch`.

**Review by whitneywhtsang — APPROVED** [review]

**Review by victor-eds — APPROVED** [review]
